### PR TITLE
segger/sysview: add syscall support

### PR DESCRIPTION
**Pull request by anchao**
## Summary

segger/sysview: add syscall support

Screenshot from 2022-03-08 11-03-49

Screenshot from 2022-03-08 11-02-41

## Impact

N/A

## Testing

ci check

### CODE REVIEW

File: drivers/segger/note_sysview.c [hunk 1]
```diff
@@ -40,12 +40,17 @@
...
-  unsigned int              irq[CONFIG_SMP_NCPUS];
+  unsigned int                 irq[CONFIG_SMP_NCPUS];
 #ifdef CONFIG_SCHED_INSTRUMENTATION_FILTER
-  struct note_filter_mode_s mode;
+  struct note_filter_mode_s    mode;
 #endif
 #ifdef CONFIG_SCHED_INSTRUMENTATION_IRQHANDLER
-  struct note_filter_irq_s  irq_mask;
+  struct note_filter_irq_s     irq_mask;
+  unsigned int                 irq_nest[CONFIG_SMP_NCPUS];
```
Comment (xiaoxiang781216) [inline drivers/segger/note_sysview.c:49]: why need record irq nest?

Comment (xiaoxiang781216) [inline drivers/segger/note_sysview.c:49]: why need record irq nest?

File: drivers/segger/note_sysview.c [hunk 2]
```diff
@@ -310,6 +399,42 @@ void sched_note_irqhandler(int irq, FAR void *handler, bool enter)
...
+    {
+      return;
+    }
+
+  if (NOTE_FILTER_SYSCALLMASK_ISSET(nr, &g_sysview.syscall_active) == 0)
+    {
+      SEGGER_SYSVIEW_NameMarker(nr, g_funcnames[nr]);
+      NOTE_FILTER_SYSCALLMASK_SET(nr, &g_sysview.syscall_active);
+
+      if (NOTE_FILTER_SYSCALLMASK_ISSET(0, &g_sysview.syscall_active) == 0)
```
Comment (xiaoxiang781216) [inline drivers/segger/note_sysview.c:415]: why?

Comment (xiaoxiang781216) [inline drivers/segger/note_sysview.c:415]: why?

File: drivers/segger/note_sysview.c [hunk 3]
```diff
@@ -310,6 +399,42 @@ void sched_note_irqhandler(int irq, FAR void *handler, bool enter)
...
+}
+
+void sched_note_syscall_leave(int nr, uintptr_t result)
+{
+  if (sysview_isenabled_syscall(nr) == 0)
+    {
+      return;
+    }
+
+  if (NOTE_FILTER_SYSCALLMASK_ISSET(nr, &g_sysview.syscall_active) != 0)
```
Comment (xiaoxiang781216) [inline drivers/segger/note_sysview.c:431]: why need check

Comment (xiaoxiang781216) [inline drivers/segger/note_sysview.c:431]: why need check

File: drivers/segger/note_sysview.c [hunk 4]
```diff
@@ -189,6 +202,82 @@ static bool sysview_isenabled_irq(int irq, bool enter)
...
+  if (!sysview_isenabled())
+    {
+      return false;
+    }
+
+  /* Exclude the case of syscall called by the interrupt handler which is
+   * not traced.
+   */
+
+  if (up_interrupt_context())
```
Comment (xiaoxiang781216) [inline drivers/segger/note_sysview.c:255]: why not trace syscall in the interrupt context?

Comment (xiaoxiang781216) [inline drivers/segger/note_sysview.c:255]: why not trace syscall in the interrupt context?

File: drivers/segger/note_sysview.c [hunk 5]
```diff
@@ -310,6 +399,42 @@ void sched_note_irqhandler(int irq, FAR void *handler, bool enter)
...
 
+#ifdef CONFIG_SCHED_INSTRUMENTATION_SYSCALL
+void sched_note_syscall_enter(int nr, int argc, ...)
+{
+  if (sysview_isenabled_syscall(nr) == 0)
+    {
+      return;
+    }
+
+  if (NOTE_FILTER_SYSCALLMASK_ISSET(nr, &g_sysview.syscall_active) == 0)
```
Comment (xiaoxiang781216) [inline drivers/segger/note_sysview.c:410]: need nr - NOTE_FILTER_SYSCALLMASK_ISSET in all NOTE_FILTER_SYSCALLMASK_xxx

Comment (xiaoxiang781216) [inline drivers/segger/note_sysview.c:410]: need nr - NOTE_FILTER_SYSCALLMASK_ISSET in all NOTE_FILTER_SYSCALLMASK_xxx

File: drivers/segger/note_sysview.c [hunk 6]
```diff
@@ -155,6 +160,14 @@ static bool sysview_isenabled(void)
     }
 #endif
 
+#ifdef CONFIG_SCHED_INSTRUMENTATION_SYSCALL
+  if (!enable &&
+      NOTE_FILTER_SYSCALLMASK_ISSET(0, &g_sysview.syscall_active))
+    {
+      NOTE_FILTER_SYSCALLMASK_ZERO(&g_sysview.syscall_active);
```
Comment (xiaoxiang781216) [inline drivers/segger/note_sysview.c:167]: why change syscall_active in sysview_isenabled?

Comment (xiaoxiang781216) [inline drivers/segger/note_sysview.c:167]: why change syscall_active in sysview_isenabled?

File: drivers/segger/note_sysview.c [hunk 7]
```diff
@@ -40,12 +40,17 @@
...
+  struct note_filter_mode_s    mode;
 #endif
 #ifdef CONFIG_SCHED_INSTRUMENTATION_IRQHANDLER
-  struct note_filter_irq_s  irq_mask;
+  struct note_filter_irq_s     irq_mask;
+  unsigned int                 irq_nest[CONFIG_SMP_NCPUS];
+#endif
+#ifdef CONFIG_SCHED_INSTRUMENTATION_SYSCALL
+  struct note_filter_syscall_s syscall_mask;
+  struct note_filter_syscall_s syscall_active;
```
Comment (xiaoxiang781216) [inline drivers/segger/note_sysview.c:53]: change syscall_active to syscall_marker

Comment (xiaoxiang781216) [inline drivers/segger/note_sysview.c:53]: change syscall_active to syscall_marker

File: drivers/segger/note_sysview.c (hunk 1 above)
Comment (anchao) [reply drivers/segger/note_sysview.c]: Done

File: drivers/segger/note_sysview.c (hunk 2 above)
Comment (anchao) [reply drivers/segger/note_sysview.c:394]: Use the Syscall "0" to identify whether the syscall is enabled, if the host tool is closed abnormally, use this bit to clear the active set.

Comment (anchao) [reply drivers/segger/note_sysview.c:394]: Use the Syscall "0" to identify whether the syscall is enabled, if the host tool is closed abnormally, use this bit to clear the active set.

File: drivers/segger/note_sysview.c (hunk 4 above)
Comment (anchao) [reply drivers/segger/note_sysview.c]: Done, removed

File: drivers/segger/note_sysview.c (hunk 5 above)
Comment (anchao) [reply drivers/segger/note_sysview.c:379]: Use the Syscall "0" to identify whether the syscall is enabled, if the host tool is closed abnormally, use this bit to clear the active set.

File: drivers/segger/note_sysview.c (hunk 7 above)
Comment (anchao) [reply drivers/segger/note_sysview.c]: Done

File: drivers/segger/note_sysview.c (hunk 1 above)
Comment (anchao) [reply drivers/segger/note_sysview.c]: Done removed

File: drivers/segger/note_sysview.c [hunk 8]
```diff
@@ -197,6 +214,45 @@ static bool sysview_isenabled_irq(int irq, bool enter)
...
+  if (!sysview_isenabled())
+    {
+      return false;
+    }
+
+  /* If the syscall trace is disabled or the syscall number is masked,
+   * do nothing.
+   */
+
+  if (!(g_sysview.mode.flag & NOTE_FILTER_MODE_FLAG_SYSCALL) ||
```
Comment (pkarashchenko) [inline drivers/segger/note_sysview.c:244]:
```suggestion
  if ((g_sysview.mode.flag & NOTE_FILTER_MODE_FLAG_SYSCALL) == 0 ||
```

File: drivers/segger/note_sysview.c [hunk 9]
```diff
@@ -490,4 +594,50 @@ void sched_note_filter_irq(struct note_filter_irq_s *oldf,
...
+ *
+ ****************************************************************************/
+
+#ifdef CONFIG_SCHED_INSTRUMENTATION_SYSCALL
+void sched_note_filter_syscall(struct note_filter_syscall_s *oldf,
+                               struct note_filter_syscall_s *newf)
+{
+  irqstate_t flags;
+
+  flags = spin_lock_irqsave(NULL);
```
Comment (pkarashchenko) [inline drivers/segger/note_sysview.c:623]: Should we introduce a separate `segger_spinlock` instance?

File: drivers/segger/note_sysview.c [hunk 10]
```diff
@@ -155,11 +159,24 @@ static bool sysview_isenabled(void)
     }
 #endif
 
+#ifdef CONFIG_SCHED_INSTRUMENTATION_SYSCALL
+  /* Use the Syscall "0" to identify whether the syscall is enabled,
```
Comment (xiaoxiang781216) [inline drivers/segger/note_sysview.c:163]: but how to handle if nobody call syscall#0?

File: drivers/segger/note_sysview.c [hunk 11]
```diff
@@ -310,6 +366,54 @@ void sched_note_irqhandler(int irq, FAR void *handler, bool enter)
 }
 #endif
 
+#ifdef CONFIG_SCHED_INSTRUMENTATION_SYSCALL
+void sched_note_syscall_enter(int nr, int argc, ...)
+{
+  if (sysview_isenabled_syscall(nr) == 0)
```
Comment (xiaoxiang781216) [inline drivers/segger/note_sysview.c:372]: let's sub nr by CONFIG_SYS_RESERVED at the beginning to ensure all place use the correct value.

File: drivers/segger/note_sysview.c [hunk 12]
```diff
@@ -310,6 +366,54 @@ void sched_note_irqhandler(int irq, FAR void *handler, bool enter)
...
+          NOTE_FILTER_SYSCALLMASK_SET(0, &g_sysview.syscall_marker);
+        }
+    }
+
+  SEGGER_SYSVIEW_MarkStart(nr);
+}
+
+void sched_note_syscall_leave(int nr, uintptr_t result)
+{
+  if (sysview_isenabled_syscall(nr) == 0)
```
Comment (xiaoxiang781216) [inline drivers/segger/note_sysview.c:405]: ditto for nr

File: boards/sim/sim/sim/configs/segger/defconfig [hunk 13]
```diff
@@ -62,6 +62,7 @@ CONFIG_SCHED_HAVE_PARENT=y
 CONFIG_SCHED_HPWORK=y
 CONFIG_SCHED_INSTRUMENTATION=y
 CONFIG_SCHED_INSTRUMENTATION_FILTER=y
+CONFIG_SCHED_INSTRUMENTATION_SYSCALL=y
```
Comment (xiaoxiang781216) [inline boards/sim/sim/sim/configs/segger/defconfig:65]: let's modify https://github.com/apache/incubator-nuttx/blob/master/tools/ci/testlist/sim-02.dat to remove segger from macOS build.

File: drivers/segger/note_sysview.c (hunk 8 above)
Comment (anchao) [reply drivers/segger/note_sysview.c:244]: Done

File: drivers/segger/note_sysview.c (hunk 9 above)
Comment (anchao) [reply drivers/segger/note_sysview.c:623]: we will plan to unified filter to common code on next PR, which will be a separate patch

File: drivers/segger/note_sysview.c (hunk 11 above)
Comment (anchao) [reply drivers/segger/note_sysview.c:373]: Done

File: boards/sim/sim/sim/configs/segger/defconfig (hunk 13 above)
Comment (anchao) [reply boards/sim/sim/sim/configs/segger/defconfig:65]: Done

File: drivers/segger/note_sysview.c (hunk 10 above)
Comment (anchao) [reply drivers/segger/note_sysview.c:163]: The 0th syscall is a special symbol (_exit()), most calls focus only on exit()

https://github.com/apache/incubator-nuttx/blob/master/include/sys/syscall_lookup.h#L27

File: tools/ci/testlist/sim-02.dat [hunk 14]
```diff
@@ -8,3 +8,7 @@
 
 # macOS doesn't have X11
 -Darwin,sim:touchscreen
+
+# macOS doesn't support --wrap flag
```
Comment (xiaoxiang781216) [inline tools/ci/testlist/sim-02.dat:12]: move before line 9 to keep the order

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**